src/meepgeom.cpp

```diff
@@ -2853,188 +2853,179 @@ void material_grids_addgradient_point(double *v, vector3 p, double scalegrad, ge
-        std::complex<double> adj = GET_FIELDS(fields_a, ci_adjoint, f_i, idx_fields);
-        material_type md;
-        geps->get_material_pt(md, p);
-        if (!md->trivial) adj *= cond_cmp(adjoint_c,p,frequencies[f_i], geps);
-        double cyl_scale;
-        int ci_forward = 0;
-        FOR_MY_COMPONENTS(forward_c) {
-          /* we need to calculate the bounds of
-          the forward fields (in space) so that we
-          can properly index into the fields array
-          later */
-          meep::ivec isf = is[ci_forward];
-          meep::ivec ief = ie[ci_forward];
-          // the actual starting index, as if we were running another LOOP_OVER_IVECS
-          ptrdiff_t idx0_f = (isf - (gv).little_corner()).yucky_val(0) / 2 * loop_s1 + \
-                             (isf - (gv).little_corner()).yucky_val(1) / 2 * loop_s2 + \
-                             (isf - (gv).little_corner()).yucky_val(2) / 2 * loop_s3;
-          start_ivec = gv.iloc(forward_c,idx0_f);
-          ptrdiff_t the_stride[5];
-          set_strides(gv.dim, the_stride, isf,ief);
-          /**************************************/
-          /*            Main Routine            */
-          /**************************************/
-          /********* compute -λᵀAᵤx *************/
-
-          /* trivial case, no interpolation/restriction needed        */
-          if (forward_c == adjoint_c) {
-            std::complex<double> fwd = GET_FIELDS(fields_f, ci_forward, f_i, idx_fields);
-            cyl_scale = (gv.dim == meep::Dcyl) ? 2*p.r() : 1; // the pi is already factored in near2far.cpp
-            material_grids_addgradient_point(
-                v+ng*f_i, vec_to_vector3(p), scalegrad*cyl_scale, geps,
+  for (size_t f_i = 0; f_i < nf; f_i++) {
+    
+    // loop over adjoint components
+    for (int ci_adjoint=0; ci_adjoint<3; ci_adjoint++){
+      int num_chunks = adjoint_dft_chunks[ci_adjoint].size();
+      if (num_chunks == 0) continue;
+      
+      // loop over each chunk
+      for (int cur_chunk=0;cur_chunk<num_chunks;cur_chunk++){
+        meep::dft_chunk* adj_chunk = adjoint_dft_chunks[ci_adjoint][cur_chunk];
+        meep::component adjoint_c = adj_chunk->c;
+        meep::grid_volume gv_adj = gv.subvolume(adj_chunk->is,adj_chunk->ie,adjoint_c);
+
+        // loop over forward components
+        for (int ci_forward=0; ci_forward<3; ci_forward++){
+          size_t num_f_chunks = forward_dft_chunks[ci_forward].size();
+          if ((num_f_chunks == 0) || (cur_chunk>=num_f_chunks)) continue;
+          meep::dft_chunk* fwd_chunk = forward_dft_chunks[ci_forward][cur_chunk];
+          meep::component forward_c = fwd_chunk->c;
+          meep::grid_volume gv_fwd = gv.subvolume(fwd_chunk->is,fwd_chunk->ie,forward_c);
+          
+          // loop over each point of interest
+          LOOP_OVER_IVECS(gv,adj_chunk->is_old,adj_chunk->ie_old,idx){
+            meep::ivec ip = gv.iloc(adjoint_c,idx);
```

Comment:
```suggestion
            IVEC_LOOP_ILOC(gv, ip);
```

Reply:
Probably need to change https://github.com/NanoComp/meep/blob/9c5582ed99251bec706400d3639f04a347d985fb/src/meep/vec.hpp#L352 to use `meep::ivec`